\documentclass[10pt,a4paper]{article}
\usepackage[papersize={210mm,297mm}, nohead, nofoot, top=10mm, bottom=10mm, outer=10mm, inner=10mm]{geometry}

% \usepackage[T2A]{fontenc}
% \usepackage[utf8]{inputenc}
\usepackage[english]{babel}

% Useful packages
\usepackage{amsmath}
\usepackage{color}
\usepackage{url}

\title{
ChiTaRS-${}_{3.1}$-the enhanced chimeric transcripts and RNA-seq database matched with protein-protein interactions
}

\author{Alessandro Gorohovski, Somnath Tagore, etc...
}

\begin{document}
\maketitle

\begin{abstract}
Discovery of chimeric RNAs, which are produced by chromosomal translocations as well as 
the joining of exons from different genes by trans-splicing, has added a new level of complexity to our study and 
understanding of the transcriptome. The enhanced ChiTaRS-${}_{3.1}$ database (\url{http://chitars.md.biu.ac.il}) is designed 
to make widely accessible a wealth of mined data on chimeric RNAs, with easy-to-use analytical tools built-in.
\end{abstract}

\section{Tables}

\begin{table}[!ht]
  \caption{The major improvements and data additions in ChiTaRS-${}_{3.1}$ in comparison to ChiTaRS-${}_{2.1}$.
}
  \tabcolsep=2mm 
  \centering
  \begin{tabular}[t]{clcrc}
c0 & c1 & c2 & c3 & c4 \\
\rule{0mm}{1.5em}
00
 &
~
01
 &
02
 &
\color{green}
\scriptsize
03
 &
04
\\
\hline
\rule{0mm}{1.5em}
10
 &
11
 &
12
 &
13
 &
14
~
\\
\hline
\hline
\color{green}
\footnotesize\color{blue}
\rule{0mm}{1.5em}
20
 &
21
 &
\multicolumn{2}{c}{
22
}
 &
24
\\ \cline{3-4}
~
\\
\hline
\large\color{red}
\rule{0mm}{1.5em}
30
 &
31
 &
32
 &
33
 &
34
~
\\
\hline
  \end{tabular}
\end{table}

\begin{center}

\begin{tabbing}
00
 \=
01
 \=
02
 \=
03
 \=
04
\\
10
 \=
11
 \=
12
 \=
13
 \=
14
\\
20
 \=
21
 \=
22
 \=
23
 \=
24
\\
30
 \=
31
 \=
32
 \=
33
 \=
34
\end{tabbing}

\end{center}


\mbox{
SPECIFY VALUE NoNameII !
%%%END:
}

\end{document}
